daemon/api_sideload_n_try.go

```diff
@@ -207,6 +212,118 @@ out:
 	return AsyncResponse(nil, chg.ID())
 }
 
+// maxReadBuflen is the maximum buffer size for reading the non-file parts in the snap upload form
+const maxReadBuflen = 1024 * 1024
+
+// readForm returns a Form populated with values (for non-file parts) and file headers (for file
+// parts). The file headers contain the original file name and a path to the persisted file in
+// dirs.SnapDirBlob. If an error occurs and a non-nil Response is returned, an attempt is made
+// to remove temp files.
+func readForm(reader *multipart.Reader) (_ *Form, apiErr *apiError) {
+	maxMemory := int64(maxReadBuflen)
+	form := &Form{
+		Values:   make(map[string][]string),
+		FileRefs: make(map[string][]*FileReference),
+	}
+
+	// clean up if we're failing the request
+	defer func() {
+		if apiErr != nil {
+			form.RemoveAll()
+		}
+	}()
+
+	for {
+		part, err := reader.NextPart()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			return nil, BadRequest("cannot read POST form: %v", err)
+		}
+
+		name := part.FormName()
+		if name == "" {
+			continue
+		}
+
+		filename := part.FileName()
+		if filename == "" {
+			// non-file parts are kept in memory
+			buf := &bytes.Buffer{}
+
+			// copy one byte more than the max so we know if it exceeds the limit
+			n, err := io.CopyN(buf, part, maxMemory+1)
+			if err != nil && !errors.Is(err, io.EOF) {
+				return nil, BadRequest("cannot read form data: %v", err)
+			}
+
+			maxMemory -= n
+			if maxMemory < 0 {
```

Comment:
We don't seem to be hitting a case where maxMemory=0 here, can you add a test case for this (right now changing this to `if maxMemory <= 0` doesn't fail any test).

Reply:
Added a test for this. The second commit is because I had some issues with an io.ErrUnexpectedEOF so I added it temporarily to an if condition while playing around. Then I forgot about it and ended up committing it. The second commit just restores the condition to what it was.